### PR TITLE
Dsc 1345 1

### PR DESCRIPTION
**Pull request by shuangzhou**
This request provides the fixes for the following points in dsc-1345: pt 4, 6, 7, 8 in the upload section of the issue.

### CODE REVIEW

**Comment by terenceingram** [comment]
@JonathanShaw @AaronCMuller can you guys have a look?

**Comment by shuangzhou** [comment]
@JonathanShaw @AaronCMuller @TerenceIngram, please feel free to update the code if you can see an improvement to be gained.  I haven't got around to try the following (and won't be able to today):
  - extend the fileuploadchange listener binded to $('#fileupload') to check data() instead of checking the data in fileuploadadd listener, and if data is valid, trigger fileuploadadd, otherwise trigger fileuploadcancel

Thanks.

File: src/banjo/controllers/JobsController.java [hunk 1]
```diff
@@ -689,20 +687,49 @@ public String receiveFile(@PathVariable long id,
...
                 response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
                 return "jobs/static_file_chunk_response";
             }
         }
+        
+        if (job.getJobType() != null && "ead".equalsIgnoreCase(job.getJobType())) {
+            try (AmberSession amberSession = amberDbService.get(job.getAmberSessionId())) {
+                // check to ensure file name has valid work object id specified
+                String msg = "invalid file {} uploaded for job {}, the file name need to be {}.xml";
+                String workObjId = jobItem.getAmberWork(amberSession).getObjId();
```
Comment (AaronCMuller) [inline src/banjo/controllers/JobsController.java]: Can this throw a NoSuchObjectException?

File: src/banjo/controllers/JobsController.java [hunk 2]
```diff
@@ -689,20 +687,49 @@ public String receiveFile(@PathVariable long id,
...
+                if (workObjId != null && !filename.equalsIgnoreCase(workObjId + ".xml")) {
```
Comment (AaronCMuller) [inline src/banjo/controllers/JobsController.java]: How would workObjId be null here?

File: src/banjo/controllers/JobsController.java [hunk 3]
```diff
@@ -689,20 +687,49 @@ public String receiveFile(@PathVariable long id,
...
                 log.error("invalid file {} uploaded for job {}", filename, job.getId());
                 response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
                 return "jobs/static_file_chunk_response";
             }
         }
+        
+        if (job.getJobType() != null && "ead".equalsIgnoreCase(job.getJobType())) {
+            try (AmberSession amberSession = amberDbService.get(job.getAmberSessionId())) {
+                // check to ensure file name has valid work object id specified
+                String msg = "invalid file {} uploaded for job {}, the file name need to be {}.xml";
```
Comment (AaronCMuller) [inline src/banjo/controllers/JobsController.java]: I'd suggest moving this down to where it's actually used.

File: src/banjo/controllers/JobsController.java [hunk 4]
```diff
@@ -689,20 +687,49 @@ public String receiveFile(@PathVariable long id,
...
+                String workObjId = jobItem.getAmberWork(amberSession).getObjId();
+                if (workObjId != null && !filename.equalsIgnoreCase(workObjId + ".xml")) {
+                    log.error(msg, filename, job.getId(), workObjId);
+                    response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+                    return "jobs/static_file_chunk_response";
+                }
+                
+                // check to ensure ead xml namespace is specified
+                InputStreamReader reader = new InputStreamReader(file.getInputStream());
```
Comment (AaronCMuller) [inline src/banjo/controllers/JobsController.java]: I'm guessing this should probably be buffered?

File: src/banjo/controllers/JobsController.java [hunk 5]
```diff
@@ -689,20 +687,49 @@ public String receiveFile(@PathVariable long id,
...
+                char[] buf = new char[256];
+                int read = reader.read(buf);
+                reader.close();
+                msg = "invalid EAD file, it is missing the EAD xml namespace: {}.";
```
Comment (AaronCMuller) [inline src/banjo/controllers/JobsController.java]: Maybe change this error message to "Invalid EAD file {}.xml. File does not contain the EAD xml namespace declaration.

File: src/banjo/models/CheckEADResponse.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package banjo.models;
+
+public class CheckEADResponse {
```
Comment (AaronCMuller) [inline src/banjo/models/CheckEADResponse.java]: I don't understand the point of this class. Could you please explain?

File: web/WEB-INF/views/jobs/eadUpload/edit_and_upload.ftl [hunk 7]
```diff
@@ -171,44 +171,14 @@
...
-<div class="modal fade" id="invalidEADFile" tabindex="-1" role="dialog" aria-labelledby="invalidEADFile" aria-hidden="true">
+<div class="modal fade" id="ingestError" tabindex="-1" role="dialog" aria-labelledby="ingestError" aria-hidden="true">
         <div class="modal-dialog">
             <div class="modal-content">
                 <div class="modal-header">
                     <button type="button" class="close" data-dismiss="modal" aria-hidden="true">&times;</button>
                 </div>    
                 <div class="modal-body">
-                    <p id="error-message">Invalid EAD file, it is missing the EAD xml namespace: ${xmlHeader}.</p>
+                    <p id="errorMessage">Invalid file name, file name should be ${work.objId}.xml.</p>
```
Comment (AaronCMuller) [inline web/WEB-INF/views/jobs/eadUpload/edit_and_upload.ftl]: If you're going to set the contents anyway, then I would suggest leaving it blank.

File: web/WEB-INF/views/jobs/eadUpload/edit_and_upload.ftl [hunk 8]
```diff
@@ -224,39 +194,124 @@ var bulkUpload=false;
...
+        $.ajax({
+		  url: 'jobs/' + jobId + '/check-eaduploaded',
+		  dataType: "json",
+		  success:function(data1){
+			  if(data1.eadUploaded){
+				  aleradyUploaded = true; 
+				  $('.alert').html(ingestErrMsg);
+				  $('.alert').attr("style", "display:block");
+				  $('.progress').html('<div id="total" class="progress-bar" aria-valuemax="100" aria-valuemin="0" aria-valuenow="45" role="progressbar"></div>');
+                  // $('.media').remove();
```
Comment (AaronCMuller) [inline web/WEB-INF/views/jobs/eadUpload/edit_and_upload.ftl]: please remove these comments.

File: web/WEB-INF/views/jobs/eadUpload/edit_and_upload.ftl [hunk 9]
```diff
@@ -224,39 +194,124 @@ var bulkUpload=false;
...
+  alreadyUploadedMsg: "EAD file already uploaded.",
+  checkAlreadyUploaded: function(data) {    
+        var alreadyUploaded = false;
+        $.ajax({
+		  url: 'jobs/' + jobId + '/check-eaduploaded',
+		  dataType: "json",
+		  success:function(data1){
+			  if(data1.eadUploaded){
+				  aleradyUploaded = true; 
+				  $('.alert').html(ingestErrMsg);
```
Comment (AaronCMuller) [inline web/WEB-INF/views/jobs/eadUpload/edit_and_upload.ftl]: What is ingestErrMsg here? If it's "failed to complete job as either..." as above, this is probably misleading.

File: web/WEB-INF/views/jobs/eadUpload/edit_and_upload.ftl [hunk 10]
```diff
@@ -224,39 +194,124 @@ var bulkUpload=false;
...
+		      $('.alert').html("Error: Status: "+textStatus+" Message: "+errorThrown);
+		      $('.alert').attr("style", "display:block");
+		      $('#errorMessage').html(ingestErrMsg);
+		  }
+		});
+		return alreadyUploaded;
+  },
   checkData: function(data) {
+  		// reset previous alerts
+  		if ($('.alert').length > 1) {
```
Comment (AaronCMuller) [inline web/WEB-INF/views/jobs/eadUpload/edit_and_upload.ftl]: should this be !=0 ?

File: web/WEB-INF/views/jobs/eadUpload/edit_and_upload.ftl [hunk 11]
```diff
@@ -224,39 +194,124 @@ var bulkUpload=false;
...
+            $('.alert').slice(1).remove();
```
Comment (AaronCMuller) [inline web/WEB-INF/views/jobs/eadUpload/edit_and_upload.ftl]: Are you trying to remove all previous alerts? If so, then why not just do $('.alert').html(""):

File: web/WEB-INF/views/jobs/eadUpload/edit_and_upload.ftl [hunk 12]
```diff
@@ -224,39 +194,124 @@ var bulkUpload=false;
...
             var reader = new FileReader();
-            reader.readAsText(blob);
+            var fileHeader;
             reader.onload = function(e) {
-                var fileHeader = reader.result;
+                fileHeader = reader.result;
                 if (!fileHeader.contains(xmlHeader)) {
-                    invalidXMLFileModal.modal('show');
-                    return false;
+                   ingestErrMsg = "Invalid EAD file, it is missing the EAD xml namespace: ${xmlHeader}.";
```
Comment (AaronCMuller) [inline web/WEB-INF/views/jobs/eadUpload/edit_and_upload.ftl]: Change error message as commented above.

File: web/assets/js/banjo-uploading.js [hunk 13]
```diff
@@ -3,7 +3,28 @@
 $(function() {
   'use strict';
 
+  var goOn = {
+	  upload: true,	  
+	  checkData: function(data) {
```
Comment (AaronCMuller) [inline web/assets/js/banjo-uploading.js]: Does this really need to do anything if it's being overridden? Also, what is goOn?

File: web/WEB-INF/views/jobs/eadUpload/edit_and_upload.ftl [hunk 14]
```diff
@@ -224,39 +194,124 @@ var bulkUpload=false;
 var acceptedFileTypes =${acceptedFileTypes};
 var workObjId="${work.objId}";
 var xmlHeader= String.replace("${xmlHeader}",/&quot;/g, '"');
```
Comment (JonathanShaw) [inline web/WEB-INF/views/jobs/eadUpload/edit_and_upload.ftl]: I think this is going to fail on non-firefox browsers, as the String generic methods are not part of the ECMAScript standard.  I think using something like 
`var xmlHeader= "${xmlHeader}".replace(/&quot;/g, '"');`  
would work more widely

Comment (JonathanShaw) [inline web/WEB-INF/views/jobs/eadUpload/edit_and_upload.ftl]: Also the calls to .contains work on firefox only. to be cross compatible we need to use indexOf

**Pull request closed without merging**